unified_planning/engines/compilers/grounder.py

```diff
@@ -327,7 +327,7 @@ def _compile(
         for qm in problem.quality_metrics:
             if isinstance(qm, MinimizeActionCosts):
                 simplifier = grounder_helper.simplifier
-                new_costs: Dict[Action, Optional[FNode]] = {}
+                new_costs: Dict["up.model.Action", Optional["up.model.Expression"]] = {}
```

Comment:
Is this change due to a mypy problem? Because it should be correct also in the previous version.

Reply:
Yes. 
Mypy, inside the `Dict` typing, does not recognize an `FNode` as an `Expression`. It has something to do with `Dict` being an invariant type. 

Maybe if we use `Mapping` instead of `Dict` this problem should be solved, but I haven't actually tried. Otherwise we have to type it to be the exact type requested (actually, same issue is with the `Substituter`).